File: Java/Search-in-Rotated-Sorted-Array.java
public class Solution {
    /**
     *@param A : an integer rotated sorted array
     *@param target :  an integer to be searched
     *return : an integer
     */
    public int search(int[] A, int target) {
        // write your code here
        if (A.length == 0) return -1;
        int left = 0, right = A.length - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] < A[right]) {
                if (A[mid] < target && target <= A[right]) left = mid;
                else right = mid;
            } else if (A[mid] > A[left]) {
                if (A[mid] > target && target >= A[left]) right = mid;
                else left = mid;
            }
        }
        if (A[left] == target) return left;
        else if (A[right] == target) return right;
        else return -1;
    }
}

File: Cpp/Search-in-Rotated-Sorted-Array.cpp
#include <vector>

class Solution {
    /**
     * param A : an integer ratated sorted array
     * param target :  an integer to be searched
     * return : an integer
     */
public:
    int search(vector<int> &A, int target) {
        // write your code here
        if (A.empty()) return -1;
        int left = 0, right = A.size() - 1, mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (A[mid] == target) return mid;
            if (A[mid] < A[right]) {
                if (target > A[mid] && target <= A[right]) left = mid + 1;
                else right = mid - 1;
            } else if (A[mid] > A[left]) {
                if (target >= A[left] && target < A[mid]) right = mid - 1;
                else left = mid + 1;
            } else {
                left ++;
            }
        }
        return -1;
    }
};
